netket/utils/group/_group.py

```diff
@@ -263,50 +263,80 @@ def character_table_readable(self) -> Tuple[List[str], Array]:
+            # random vector s and all regular projectors P
+            # These are calculated as linear combinations of sᴴρv for the
+            # regular rep matrices ρ, which is given by the latter two terms
+            vs = v[:, starting_idx]
+            s = np.random.standard_normal(len(self))
+            s = s[self.product_table[np.ix_(self.inverse, self.inverse)]]
+            proj = self.character_table().conj() @ s @ vs
+            starting_idx = list(starting_idx) + [len(self)]
+            return v, starting_idx, proj
+
+        eigen = {}
+        if np.any(frob == 1):
+            # real irreps: start from a real symmetric invariant matrix
+            e = np.random.standard_normal(len(self))
+            eigen["real"] = invariant_subspaces(e)
+        if np.any(frob != 1):
+            # complex or quaternionic irreps: complex hermitian invariant matrix
+            e = np.random.standard_normal((2, len(self)))
+            e = e[0] + 1j * e[1]
+            eigen["cplx"] = invariant_subspaces(e)
 
         irreps = []
 
-        for chi in self.character_table():
-            # Check which eigenvectors belong to this irrep
-            # the last argument of einsum is the regular projector on this irrep
-            proj = np.einsum(
-                "gi,hi,gh->i", vs.conj(), vs, chi.conj()[self.product_table]
-            )
-            proj = np.logical_not(np.isclose(proj, 0.0))
-            # Pick the first eigenvector in this irrep
-            idx = np.arange(vs.shape[1], dtype=int)[proj][0]
-            v = vs[:, idx]
-            # Generate an orthonormal basis for the irrep spanned by A_reg(g).v
-            # for all matrices in the regular representation
-            # A basis follows (with probability 1) as d random linear
-            # combinations of these vectors; make them orthonormal using QR
-            # NB v[product_table] generates a matrix whose columns are A_reg(h).v
-            dim = int(np.rint(chi[0].real))
-            w, _ = np.linalg.qr(v[true_product_table] @ random((len(self), dim)))
-            # Project the regular representation on this basis
+        for i, chi in enumerate(self.character_table()):
+            v, idx, proj = eigen["real"] if frob[i] == 1 else eigen["cplx"]
+            # Check which eigenspaces belong to this irrep
+            proj = np.logical_not(np.isclose(proj[i], 0.0))
+            # Pick the first eigenspace in this irrep
+            first = np.arange(len(idx) - 1, dtype=int)[proj][0]
+            v = v[:, idx[first] : idx[first + 1]]
+            # v is now the basis of a single irrep: project the regular rep onto it
             irreps.append(
                 prune_zeros(
-                    np.einsum("gi,ghj ->hij", w.conj(), w[true_product_table, :])
+                    np.einsum("gi,ghj ->hij", v.conj(), v[true_product_table, :])
                 )
             )
 
         return irreps
```

Comment:
Can you make sure the dtype returned is a float if the irreps are real and otherwise complex? Just to make sure we don't end up in situations where complex numbers are used when they don't need to be

Reply:
Nevermind, I looked more closely and this should be the case, carry on